Handle promotion failures, deadlocks, and state transitions

ralcolea opened the pull request:
This PR introduces several reliability and observability improvements to uRaft module, addressing multiple edge cases around promotion/demotion handling, deadlock prevention, and cluster recovery.

The main motivation was caused by issue #576 (promotion deadlocks), but there were introduced other improvements in the high-availability mechanism like logging and fixes when handling overlapping transitions or loss of quorum.

**Key Changes**

- **Improved handling of promotion timeouts and failures**
   - Introduces `handlePromotionFailure()` and `cleanupDirtyPromotion()` in `uRaftController`.
   - Automatically triggers cleanup of stale temporary metadata file (`metadata.sfs.tmp`).
   - Ensures the node can rejoin as a shadow after failed or interrupted promotions.
   - Floating IP manager now starts only after a successful promotion.

- **Removed redundant `set_block_promotion()` calls**
   - Prevented unnecessary election blocking in saunafs-uraft-helper and controller logic.
   - Resolved potential promotion deadlocks caused by overlapping state blocking.

- **Logging and Observability**
   - Added structured syslog messages for:
      - Promotion, demotion, and quorum-related transitions.
      - Manual demotions and heartbeat-based state changes.
      - Election victories and higher-term detections.

   - Introduced `nodeToString()` helper for consistent and readable node identifiers in logs.
   - Enhanced logs in `saunafs_drop_ip()` for clarity during IP removal and state cleanup.

- **Deferred promotions during active demotion**
   - Added `force_promote_` flag to defer a promotion request until demotion completes.
   - Ensures that overlapping commands no longer cause promotion loss or inconsistent states.
   - Logs forced transitions for traceability.

**Testing**

The following Molecule scenarios were added and refactored to test the new changes:
- `uraft_kill_sfsmaster:` This scenario was refactored to restart uRaft in the server where the `sfsmaster` process is killed, ensuring the cluster finishes in a stable state.
- `uraft_loss_quorum:` This scenario was modified to simulate a quorum loss and validate that the cluster can continue serving I/O operations once the quorum is restored.
- `uraft_deadlock_multiple_elections:` This is a new scenario introduced to validate that the uRaft cluster remains stable and free of deadlocks during consecutive election cycles.
- `uraft_failed_promotion:` This is a new scenario introduced to verify that the uRaft cluster can gracefully recover when a promotion attempt fails. It also ensures that `handlePromotionFailure()` works as expected in case of failed promotions.

For more details, see `fix-saunafs-issue-576` branch in deployment (Ansible) repository.

ralcolea committed:
fix: Handle failure promotions

This commit applies the following changes intended to fix issue #576:
- Improve states transition during promotion by ensuring that after
  starting one demotion/promotion operation, it is not interrupted
  by other demotion/promotion operation.

- Trigger cleanup after promotion timeout to ensure a temporary master
  can be restarted as a shadow after an interrupted metadata dump or
  a failed promotion that creates a remanent metadata.sfs.tmp file that
  prevent a master restart.

- Fix Floating IP timing to ensure the manager is only started after
  a successful promotion.

Signed-off-by: Crash <crash@leil.io>

ralcolea committed:
fix(uraft): Prevent deadlock on blocked promotions

Previously, certain execution paths could permanently block elections
by calling `set_block_promotion(true)`, leading to a deadlock scenario
where no Leader could be elected. Some of these calls were unnecessary,
as the uRaft mechanism based on the loyalty agreement is already
designed to prevent election loops correctly.

This commit addresses issue #576 by:
- Removing redundant calls to `set_block_promotion()` to avoid the
  deadlock scenario described above.
- Improving logging in the `saunafs_drop_ip` function to provide more
  detailed information about the status of the floating IP removal.

Signed-off-by: Crash <crash@leil.io>

ralcolea committed:
refactor(uraft): Add promotion/demotion logs

This commit adds detailed logs for promotion and demotion events in
uRaft that were not previously recorded. These logs improve visibility
into node state transitions and help identify issues during debugging.

Signed-off-by: Crash <crash@leil.io>

ralcolea committed:
fix(uraft): Defer promotion request during demotion

Previously, if a promotion was requested while a demotion was still in
progress, the request was ignored. This could leave the node stuck in
a follower state even when a promotion was desired.

This commit adds a `force_promote_` flag to defer the promotion until
the current demotion completes. Once demotion finishes, the controller
automatically triggers a forced promotion and logs the transition.

This ensures consistent role transitions and prevents promotion loss
during overlapping commands.

Signed-off-by: Crash <crash@leil.io>

ralcolea commented on src/uraft/uraftcontroller.cc:
```diff
@@ -228,9 +246,20 @@ void uRaftController::checkNodeStatus(const boost::system::error_code &error) {
+                syslog(LOG_WARNING, "Demotion timeout - killing stuck command");
+            } else if (command_type_ == kCmdStatusDead) {
+                syslog(LOG_WARNING, "Dead handler timeout - killing stuck command");
+            }
```
Nice catch @copilot :+1: 
Fixed.

ralcolea commented on src/uraft/saunafs-uraft-helper.in:
```diff
@@ -292,9 +322,28 @@ saunafs_assign_ip() {
 
 saunafs_drop_ip() {
 	load_config
-	sudo ip -f inet addr delete $ipaddr/$netmask dev $iface
-	if [[ "$ipaddr2" != "" ]]; then
-		sudo ip -f inet addr delete $ipaddr2/$netmask2 dev $iface2
+
+	# Check if floating IP exists before trying to drop it
+	if is_ip_present "${ipaddr}" "${iface}"; then
+		log "Dropping floating IP: ${ipaddr} from interface: ${iface}"
+		if ! sudo ip -f inet addr delete $ipaddr/$netmask dev $iface; then
+			log "Warning: Failed to drop floating IP: ${ipaddr}"
+		else
+			log "Successfully dropped floating IP: ${ipaddr}"
+		fi
+	elif [[ "$ipaddr" != "" ]]; then
+		log "Floating IP ${ipaddr} not present - nothing to drop"
+	fi
+
+	if [[ "$ipaddr2" != "" ]] && is_ip_present "$ipaddr2" "$iface2"; then
+		log "Dropping floating IP: ${ipaddr2} from interface: ${iface2}"
+		if ! sudo ip -f inet addr delete $ipaddr2/$netmask2 dev $iface2; then
+			log "Warning: Failed to drop floating IP: ${ipaddr2}"
+		else
+			log "Successfully dropped floating IP: ${ipaddr2}"
+		fi
+	elif [[ "$ipaddr2" != "" ]]; then
+		log "Floating IP ${ipaddr2} not present - nothing to drop"
 	fi
```
Your proposal matches exactly the current version.

ralcolea commented on src/uraft/uraft.cc:
```diff
@@ -230,16 +236,21 @@ void uRaft::heartbeat(const boost::system::error_code &error) {
 				state_.type      = kFollower;
 				state_.voted_for = -1;
+				state_.current_term = 0;  // Reset term to avoid former leader isolation deadlock
+				state_.leader_id    = -1; // Reset leader ID to elect a new leader in next election
```
Your comments suggestions match exactly with the current version of the code @gemini.

ralcolea commented on src/uraft/uraftcontroller.cc:
```diff
@@ -417,3 +446,17 @@ void uRaftController::stopFloatingIpManager() {
+	std::vector<std::string> cleanup = {"saunafs-uraft-helper", "cleanup"};
+	std::string result;
+
+	if (!runCommand(cleanup, result, 5000)) {
+		syslog(LOG_ERR, "Failed to cleanup dirty state: %s", result.c_str());
+	}
```
`cleanup` command triggers a call to `saunafs_cleanup_dirty_state()` function in the `saunafs-uraft-helper` script, and this function logs whether the cleanup was succesful or not. Adding a log for the successful scenario will be redundant @gemini.

ralcolea commented on src/uraft/uraftcontroller.cc:
```diff
@@ -157,18 +152,42 @@ void uRaftController::checkCommandStatus(const boost::system::error_code &error)
 		cmd_timeout_timer_.cancel();
+
+		// Check if command completed successfully
+		bool commandSucceeded = WIFEXITED(status) && WEXITSTATUS(status) == 0;
```
Nice suggestion @gemini. However, currently there is a single occurrence for this line.
If more occurrences are needed in the future, I will consider to apply your suggestion.

ralcolea commented on src/uraft/uraftcontroller.cc:
```diff
@@ -417,3 +446,17 @@ void uRaftController::stopFloatingIpManager() {
+}
+
+void uRaftController::handlePromotionFailure() {
+	demoteLeader();
+	cleanupDirtyPromotion();
```
`demoteLeader()` function is not expected to fail because it mostly modifies variables's values and start timers.
Additionally, `cleanupDirtyPromotion()` function only removes remanent files of possible failed promotions, like `metadata.sfs.tmp` or `.sfsmaster.lock` (if not process is holding it).

rolysr commented on src/uraft/uraft.cc:
```diff
@@ -95,6 +97,8 @@ void uRaft::checkTerm(int /*id*/, const RpcHeader &data) {
 	if (data.term > state_.current_term) {
 		if (state_.president) {
 			assert(state_.type != kFollower);
+			syslog(LOG_NOTICE, "(%s): Higher term detected (%lu): Node %s starting new election",
+			       __func__, data.term, nodeToString(state_.id).c_str());
```
Consider if it would be good to start moving these prints to `safs::log_info` and similar.

ralcolea replied:
I plan to apply this change for all the `syslog()` calls, but in a different PR which will only do that.

lgsilva3087 commented on src/uraft/saunafs-uraft-helper.in:
```diff
@@ -196,6 +196,36 @@ saunafs_demote() {
+				return $SAUNAFS_URAFT_ERROR
+			}
+		fi
+	fi
```
This file should not be deleted. The new master should be able to use it if no other process is holding it.

ralcolea replied:
Applied.

lgsilva3087 commented on src/uraft/uraftcontroller.cc:
```diff
@@ -26,6 +26,7 @@ uRaftController::uRaftController(boost::asio::io_context &ios)
 	command_pid_  = -1;
 	command_type_ = kCmdNone;
 	force_demote_ = false;
+	force_promote_ = false;
```
Consider isDemotePending and isPromotePending or similar.

ralcolea replied:
You're right. The current names are not matching exactly with their functions. I will rename both 👍

lgsilva3087 commented on src/uraft/uraftcontroller.h:
```diff
@@ -75,13 +75,17 @@ class uRaftController : public uRaftStatus {
 	void startFloatingIpManager();
 	void stopFloatingIpManager();
 
+	void cleanupDirtyPromotion();
+	void handlePromotionFailure();
```
Please document them, why were introduced.

ralcolea replied:
Nice suggestion. Applied.

lgsilva3087 commented on src/uraft/uraftcontroller.h:
```diff
@@ -75,13 +75,17 @@ class uRaftController : public uRaftStatus {
 	int                         command_type_;  /// Last run command type.
 	Timer                       command_timer_;
 	bool                        force_demote_;
+	bool                        force_promote_;
```
Document, it was not trivial to understand them.

ralcolea replied:
Nice suggestion. Applied.

lgsilva3087 commented on src/uraft/saunafs-uraft-helper.in:
```diff
@@ -196,6 +196,36 @@ saunafs_demote() {
+			log "Failed to remove metadata.sfs.tmp"
+			return $SAUNAFS_URAFT_ERROR
+		}
+	fi
```
Only remove it if no process is using it. The child process in a dumping could be using it.

ralcolea replied:
Nice catch :) 
Applied.

lgsilva3087 reviewed (commented):
It is looking great. Please see my comments.

ralcolea committed:
chore: Apply review suggestions

This commit introduces the following review suggestions:
- Updated the `cleanup` command in saunafs-uraft-helper to remove the
  temporary metadata.sfs.tmp file only if it is not holded by any
  process, and removed redundant logic for deleting `.sfsmaster.lock`.

- Refactor `saunafs_drop_ip` command in saunafs-uraft-helper to make
  the code more readable.

- Renamed `force_demote_` and `force_promote_` to `is_demote_pending_`
  and `is_promote_pending_` for clearer intent and consistency with
  their behavior.

- Added documentation for newly introduced functions and attributes
  to improve code readability and maintenance.

Co-authored-by: guillex <guillex@leil.io>
Co-authored-by: Rolando Sánchez Ramos <rolysr@leil.io>
Signed-off-by: Crash <crash@leil.io>

lgsilva3087 commented on src/uraft/saunafs-uraft-helper.in:
```diff
@@ -196,6 +196,27 @@ saunafs_demote() {
+
+	log "Attempting to clean up dirty metadata state"
+
+	# Remove temporary metadata file
```
Maybe we can add to the comment that it is only created during dumping, and if no process is holding it, it could represent a crash during dumping.

ralcolea replied:
I liked your suggestion :+1: 
Applied.

lgsilva3087 reviewed (approved):
Great job. Please see my small suggestion.

rolysr commented on src/uraft/saunafs-uraft-helper.in:
```diff
@@ -331,7 +331,7 @@ saunafs_drop_ip() {
 		else
 			log "Successfully dropped floating IP: ${ipaddr}"
 		fi
-	else
+	elif [[ "$ipaddr" != "" ]]; then
 		log "Floating IP ${ipaddr} not present - nothing to drop"
 	fi
 
```
Check if this block logic could be more readable like this:
```
if [[ -n "$ipaddr" ]]; then
	if is_ip_present "$ipaddr" "$iface"; then
		log "Dropping floating IP: $ipaddr from interface: $iface"
		if ! sudo ip -f inet addr delete "$ipaddr/$netmask" dev "$iface"; then
			log "Warning: Failed to drop floating IP: $ipaddr"
		else
			log "Successfully dropped floating IP: $ipaddr"
		fi
	else
		log "Floating IP $ipaddr not present - nothing to drop"
	fi
fi
```
This is for being able to not even check the `is_ip_present` if the `ipaddr` is empty. Consider the same for `ipaddr2`

ralcolea replied:
Nice suggestion. Applied :+1:

rolysr commented on src/uraft/uraft.cc:
```diff
@@ -507,3 +522,18 @@ int uRaft::scanLocalInterfaces() {
 }
+
+std::string uRaft::nodeToString(int id) {
+	if (id < 0 || id >= (int)opt_.server.size()) {
```
Consider using the std::static_cast<int>(opt_.server.size()) for a more C++-styled code.

ralcolea replied:
This code was moved exactly from another function, and despite this suggestion is out of the scope of this PR, I applied it.

rolysr reviewed (approved):
Great job @ralcolea, consider my minor suggestions.
